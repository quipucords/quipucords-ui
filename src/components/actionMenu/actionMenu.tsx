/**
 * A generic dropdown menu component for actions, utilizing PatternFly components. It allows for actions to be specified
 * per item, with each action having a label and an onClick handler. This component is flexible for use with various
 * item types and action configurations.
 *
 * @module actionMenu
 */
import React, { useState } from 'react';
import { Dropdown, DropdownItem, DropdownList, MenuToggle, type MenuToggleElement } from '@patternfly/react-core';
import { EllipsisVIcon } from '@patternfly/react-icons';

type Action<T> = {
  label: string;
  onClick: (item: T) => void;
  disabled?: boolean;
  ouiaId?: string; // Optional ouiaId for E2E testing
};

interface ActionMenuProps<T = unknown> {
  item: T;
  actions: Action<T>[];
}

const ActionMenu = <T,>({ item, actions }: ActionMenuProps<T>) => {
  const [isOpen, setIsOpen] = useState<boolean>(false);

  return (
    <Dropdown
      isOpen={isOpen}
      onOpenChange={setIsOpen}
      onSelect={() => setIsOpen(false)}
      toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
        <MenuToggle
          ref={toggleRef}
          aria-label="Action Menu Toggle"
          variant="plain"
          onClick={() => setIsOpen(prev => !prev)}
          isExpanded={isOpen}
        >
          <EllipsisVIcon />
        </MenuToggle>
      )}
      shouldFocusToggleOnSelect
    >
      <DropdownList>
        {actions.map(a => (
          <DropdownItem
            value={a.label}
            key={a.label}
            ouiaId={a.ouiaId}
            onClick={() => {
              a.onClick(item);
            }}
            isDisabled={a.disabled}
          >
            {a.label}
          </DropdownItem>
        ))}
      </DropdownList>
    </Dropdown>
  );
};

export { ActionMenu as default, ActionMenu, type ActionMenuProps };
