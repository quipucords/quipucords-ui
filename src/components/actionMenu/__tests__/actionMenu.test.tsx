import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { shallowComponent } from '../../../../config/jest.setupTests';
import { ActionMenu } from '../actionMenu';

describe('ActionMenu', () => {
  it('should render a basic component', async () => {
    const props = {
      item: { foo: 'bar' },
      actions: [{ label: 'Lorem ipsum', onClick: item => jest.fn(item) }]
    };
    const component = await shallowComponent(<ActionMenu {...props} />);
    expect(component).toMatchSnapshot('basic');
  });

  it('should display items and allow click after toggle is clicked', async () => {
    const user = userEvent.setup();
    const onMockClick = jest.fn();

    const props = {
      item: { foo: 'bar' },
      actions: [{ label: 'Lorem ipsum', onClick: onMockClick }]
    };

    const { asFragment } = render(<ActionMenu {...props} />);

    await user.click(screen.getByRole('button'));
    expect(asFragment()).toMatchSnapshot('expanded');

    await user.click(screen.getByText('Lorem ipsum'));
    expect(onMockClick.mock.calls).toMatchSnapshot('click');
  });

  it('should display tooltip', async () => {
    const user = userEvent.setup();
    const onMockClick = jest.fn();
    const tooltipProps = { content: 'dolor sit' };

    const props = {
      item: { foo: 'bar' },
      actions: [{ label: 'Lorem ipsum', onClick: onMockClick, tooltipProps: tooltipProps }]
    };

    render(<ActionMenu {...props} />);

    await user.click(screen.getByRole('button'));
    await user.hover(screen.getByText('Lorem ipsum'));
    expect(document.body).toMatchSnapshot('hover');
  });

  it('should not allow to click disabled item', async () => {
    const user = userEvent.setup();
    const onMockClick = jest.fn();

    const props = {
      item: { foo: 'bar' },
      actions: [{ label: 'Lorem ipsum', onClick: onMockClick, disabled: true }]
    };

    render(<ActionMenu {...props} />);

    await user.click(screen.getByRole('button'));
    await user.click(screen.getByText('Lorem ipsum'));
    expect(onMockClick).not.toHaveBeenCalled();
  });
});
