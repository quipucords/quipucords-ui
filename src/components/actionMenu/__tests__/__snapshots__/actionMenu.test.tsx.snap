// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders proper dom including elipse icon 1`] = `
<DocumentFragment>
  <button
    aria-expanded="false"
    aria-label="Action Menu Toggle"
    class="pf-v5-c-menu-toggle pf-m-plain"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="pf-v5-svg"
      fill="currentColor"
      height="1em"
      role="img"
      viewBox="0 0 192 512"
      width="1em"
    >
      <path
        d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`snapshot matches after opening the action menu 1`] = `
<DocumentFragment>
  <button
    aria-expanded="true"
    aria-label="Action Menu Toggle"
    class="pf-v5-c-menu-toggle pf-m-expanded pf-m-plain"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="pf-v5-svg"
      fill="currentColor"
      height="1em"
      role="img"
      viewBox="0 0 192 512"
      width="1em"
    >
      <path
        d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
      />
    </svg>
  </button>
  <div
    class="pf-v5-c-menu"
    data-ouia-component-id="OUIA-Generated-Dropdown-4"
    data-ouia-component-type="PF5/Dropdown"
    data-ouia-safe="true"
    data-popper-escaped="true"
    data-popper-placement="bottom-start"
    data-popper-reference-hidden="true"
    style="position: absolute; left: 0px; top: 0px; z-index: 9999; opacity: 1; transition: opacity 0ms cubic-bezier(.54, 1.5, .38, 1.11); min-width: 0px; transform: translate(0px, 0px);"
  >
    <div
      class="pf-v5-c-menu__content"
    >
      <ul
        class="pf-v5-c-menu__list"
        role="menu"
      >
        <li
          class="pf-v5-c-menu__list-item"
          data-ouia-component-id="OUIA-Generated-DropdownItem-2"
          data-ouia-component-type="PF5/DropdownItem"
          data-ouia-safe="true"
          role="none"
        >
          <button
            class="pf-v5-c-menu__item"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            <span
              class="pf-v5-c-menu__item-main"
            >
              <span
                class="pf-v5-c-menu__item-text"
              >
                Delete
              </span>
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
